[BE] Enable ZeRO test on windows

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65519
* __->__ #65385

Enables the ZeRO tests to run on windows. Closes
https://github.com/pytorch/pytorch/issues/63086.

Backend == NCCL was used as a proxy to see if we were running under CUDA, but Windows GPU tests uses Gloo. In this case use Gloo on GPU.

For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.

Differential Revision: [D31071181](https://our.internmc.facebook.com/intern/diff/D31071181/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

rohan-varma committed:
[BE] Enable ZeRO test on windows

Enables the ZeRO tests to run on windows. Closes
https://github.com/pytorch/pytorch/issues/63086.

Backend == NCCL was used as a proxy to see if we were running under CUDA, but Windows GPU tests uses Gloo. In this case use Gloo on GPU.

For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.

Differential Revision: [D31071181](https://our.internmc.facebook.com/intern/diff/D31071181/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE] Enable ZeRO test on windows"


Enables the ZeRO tests to run on windows. Closes
https://github.com/pytorch/pytorch/issues/63086.

Backend == NCCL was used as a proxy to see if we were running under CUDA, but Windows GPU tests uses Gloo. In this case use Gloo on GPU.

For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.

Differential Revision: [D31071181](https://our.internmc.facebook.com/intern/diff/D31071181/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE] Enable ZeRO test on windows"


Enables the ZeRO tests to run on windows. Closes
https://github.com/pytorch/pytorch/issues/63086.

Backend == NCCL was used as a proxy to see if we were running under CUDA, but Windows GPU tests uses Gloo. In this case use Gloo on GPU.

For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.

Differential Revision: [D31071181](https://our.internmc.facebook.com/intern/diff/D31071181/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE] Enable ZeRO test on windows"


Enables the ZeRO tests to run on windows. Closes
https://github.com/pytorch/pytorch/issues/63086.

Backend == NCCL was used as a proxy to see if we were running under CUDA, but Windows GPU tests uses Gloo. In this case use Gloo on GPU.

For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.

Differential Revision: [D31071181](https://our.internmc.facebook.com/intern/diff/D31071181/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

[ghstack-poisoned]

mrshenli commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -33,24 +33,27 @@
 DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
 
 
+
```
unintentional new line?

mrshenli commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -33,24 +33,27 @@
-    dist.Backend.NCCL if not IS_WINDOWS and torch.cuda.is_available()
-    else dist.Backend.GLOO
-)
+def _get_backend_for_tests():
```
curious, is this a better convention than the existing code?

rohan-varma replied:
I mainly needed it for L686, it would look strange if we had `backend = BACKEND if use_gpu else GLOO`, so went with this instead.

mrshenli reviewed (commented):
LGTM! Thanks for making it work!

> For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.

Just to confirm I understand this correctly. This means ZeRO test on Windows only runs CPU for now?

rohan-varma commented:
> LGTM! Thanks for making it work!
> 
> > For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.
> 
> Just to confirm I understand this correctly. This means ZeRO test on Windows only runs CPU for now?

The zero tests are mainly all GPU, except for one, `test_zero_join_cpu`, so Windows runs GPU tests as well but using Gloo. For other OS, backend used is NCCL for all GPU tests. This was the original source of the error, windows was trying to run on NCCL backend which is not supported.

rohan-varma committed:
Update on "[BE] Enable ZeRO test on windows"


Enables the ZeRO tests to run on windows. Closes
https://github.com/pytorch/pytorch/issues/63086.

Backend == NCCL was used as a proxy to see if we were running under CUDA, but Windows GPU tests uses Gloo. In this case use Gloo on GPU.

For some reason these tests don't seem to test Gloo on GPU with ZeRO in general (picks NCCL backend when GPU is available), so kept that behavior for now.

Differential Revision: [D31071181](https://our.internmc.facebook.com/intern/diff/D31071181/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang gcramer23

[ghstack-poisoned]
